### PR TITLE
Revise and extend Initialization Guide

### PR DESCRIPTION
**Pull request by vishaalkapoor**
* Added content from D2L parameter initialization
* Added some further detail and explanation. Broke up a few paragraphs for easier skimming.
* Cleaned up whitespace/formatting issues.

I ran this through md2ipynb to make sure it compiled to a Notebook and looked appropriate.

Completes card https://github.com/mli/new-docs/projects/1#card-15424478

### CODE REVIEW

**Comment by vishaalkapoor** [comment]
Hi @astonzhang, would you be able to take a look? Thanks!

File: python/guide/module/gluon/init.md [hunk 1]
```diff
@@ -161,3 +170,127 @@ net.add(nn.Dense(10, in_units=256))
 
 net.initialize(init=MyInit())
 ```
+
+## Parameter Initialization
+
+Now that we know how to access the parameters, let's look at how to initialize
+them properly. We discussed the need for
+[Initialization](../chapter_deep-learning-basics/numerical-stability-and-init.md)
```
Comment (NRauschmayr) [inline python/guide/module/gluon/init.md]: The link needs to be adjusted.

File: python/guide/module/gluon/init.md [hunk 2]
```diff
@@ -161,3 +170,127 @@ net.add(nn.Dense(10, in_units=256))
...
+net[0].weight.data()[0,0] = 42
+net[0].weight.data()[0]
+```
+
+## Tied Parameters
+
+In some cases, we want to share model parameters across multiple layers. For
+instance when we want to find good word embeddings we may decide to use the
+same parameters both for encoding and decoding of words. We discussed one such
+case when we introduced [Blocks](model-construction.md). Let's see how to do
```
Comment (NRauschmayr) [inline python/guide/module/gluon/init.md]: Link needs to be adjusted

File: python/guide/module/gluon/init.md (hunk 1 above)
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: Removed sentence as it was D2L specific.

File: python/guide/module/gluon/init.md (hunk 2 above)
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: I've removed the link as there isn't a Blocks link. As mentioned in the other PR we should revisit to figure out how to link between pages. I'll add a SIM.

Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: I've removed the link as there isn't a Blocks link. As mentioned in the other PR we should revisit to figure out how to link between pages.

**Comment by vishaalkapoor** [comment]
@NRauschmayr addressed concerns, thanks!

**Comment by NRauschmayr** [comment]
LGTM

File: python/guide/module/gluon/init.md [hunk 3]
```diff
@@ -161,3 +170,127 @@ net.add(nn.Dense(10, in_units=256))
...
+print(net[1].weight.data()[0,0])
+print(net[0].weight.data()[0])
+```
+
+### Custom Initialization
+
+Sometimes, the initialization methods we need are not provided in the `init`
+module. At this point, we can implement a subclass of the `Initializer` class
+so that we can use it like any other initialization method. Usually, we only
+need to implement the `_init_weight` function and modify the incoming `NDArray`
```
Comment (thomelane) [inline python/guide/module/gluon/init.md]: would reword: modify the incoming NDArray according to the initial result.

File: python/guide/module/gluon/init.md [hunk 4]
```diff
@@ -161,3 +170,127 @@ net.add(nn.Dense(10, in_units=256))
...
+        data *= data.abs() >= 5
+
+net.initialize(MyInit(), force_reinit=True)
+net[0].weight.data()[0]
+```
+
+If this functionality is insufficient, we can even set parameters directly.
+Since `data()` returns an `NDArray` we can access it just like any other matrix.
+A note for advanced users - if you want to adjust parameters within an
+`autograd` scope you need to use `set_data` to avoid confusing the automatic
```
Comment (thomelane) [inline python/guide/module/gluon/init.md]: Can you add links to API docs for these methods?

File: python/guide/module/gluon/init.md [hunk 5]
```diff
@@ -161,3 +170,127 @@ net.add(nn.Dense(10, in_units=256))
...
+net[0].weight.data()[0,0] = 42
+net[0].weight.data()[0]
+```
+
+## Tied Parameters
+
+In some cases, we want to share model parameters across multiple layers. For
+instance when we want to find good word embeddings we may decide to use the
+same parameters both for encoding and decoding of words. We discussed one such
+case when we introduced `Blocks`. Let's see how to do
```
Comment (thomelane) [inline python/guide/module/gluon/init.md]: `Blocks`? `Block`s, and when did we discuss this? Would add link or remove.

File: python/guide/module/gluon/init.md [hunk 6]
```diff
@@ -161,3 +170,127 @@ net.add(nn.Dense(10, in_units=256))
...
+this a bit more elegantly. In the following we allocate a dense layer and then
```
Comment (thomelane) [inline python/guide/module/gluon/init.md]: allocate -> construct

File: python/guide/module/gluon/init.md [hunk 7]
```diff
@@ -161,3 +170,127 @@ net.add(nn.Dense(10, in_units=256))
...
+use its parameters specifically to set those of another layer.
+
+```{.python .input  n=14}
+net = nn.Sequential()
+# we need to give the shared layer a name such that we can reference its parameters
```
Comment (thomelane) [inline python/guide/module/gluon/init.md]: we -> We

File: python/guide/module/gluon/init.md [hunk 8]
```diff
@@ -161,3 +170,127 @@ net.add(nn.Dense(10, in_units=256))
...
+
+# Check whether the parameters are the same
+print(net[1].weight.data()[0] == net[2].weight.data()[0])
+net[1].weight.data()[0,0] = 100
+# And make sure that they're actually the same object rather than just having the same value.
+print(net[1].weight.data()[0] == net[2].weight.data()[0])
+```
+
+The above exampe shows that the parameters of the second and third layer are
+tied. They are identical rather than just being equal. That is, by changing one
```
Comment (thomelane) [inline python/guide/module/gluon/init.md]: They are identical rather than just being equal. identical vs equal?

File: python/guide/module/gluon/init.md [hunk 9]
```diff
@@ -161,3 +170,127 @@ net.add(nn.Dense(10, in_units=256))
...
+of the parameters the other one changes, too. What happens to the gradients is
+quite ingenious. Since the model parameters contain gradients, the gradients of
+the second hidden layer and the third hidden layer are accumulated in the
+`shared.params.grad( )` during backpropagation.
```
Comment (thomelane) [inline python/guide/module/gluon/init.md]: `shared.params.grad( )`: I don't think this is a method

File: python/guide/module/gluon/init.md [hunk 10]
```diff
@@ -2,17 +2,14 @@
...
-did the following things that *shouldn't* work:
-
-* We defined the network
-architecture with no regard to the input dimensionality.
-* We added layers
-without regard to the output dimension of the previous layer.
-* We even
-'initialized' these parameters without knowing how many parameters were were to
-initialize.
+In the previous examples we played fast and loose with setting up our networks.
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: referencing nn section. done

File: python/guide/module/gluon/init.md [hunk 11]
```diff
@@ -21,15 +18,17 @@ convolutional networks and images this problem will become even more pertinent,
 since the input dimensionality (i.e. the resolution of an image) will affect the
 dimensionality of subsequent layers at a long range. Hence, the ability to set
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: done

File: python/guide/module/gluon/init.md [hunk 12]
```diff
@@ -21,15 +18,17 @@ convolutional networks and images this problem will become even more pertinent,
...
 parameters without the need to know at the time of writing the code what the
-dimensionality is can greatly simplify statistical modeling. In what follows, we
-will discuss how this works using initialization as an example. After all, we
-cannot initialize variables that we don't know exist.
+dimensionality is can greatly simplify statistical modeling. In what follows,
+we will discuss how this works using initialization as an example. After all,
+we cannot initialize variables that we don't know exist.
+
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: i deleted some lines

File: python/guide/module/gluon/init.md (hunk 5 above)
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: I removed the nostalgic comment.

File: python/guide/module/gluon/init.md (hunk 8 above)
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: I think that part makes sense. Identical in the sense that the pointers are identical, not just equal. The subsequent sentence explains. Let me know what you thiknk

File: python/guide/module/gluon/init.md (hunk 9 above)
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: removed 'the'  - shared is a variable above.

**Comment by vishaalkapoor** [comment]
@thomelane Please take a look

File: python/guide/module/gluon/init.md (hunk 6 above)
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: done

File: python/guide/module/gluon/init.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: done

File: python/guide/module/gluon/init.md (hunk 9 above)
Comment (thomelane) [reply python/guide/module/gluon/init.md]: Was asking about `grad`. It's usually a property not a method.

**Comment by thomelane** [comment]
Still a few comments to be addressed. Your conflict is probably due to some rehoming for tutorials, now at `python/guide/packages` instead of `python/guide/modules`.

File: python/guide/module/gluon/init.md [hunk 13]
```diff
@@ -21,15 +18,17 @@ convolutional networks and images this problem will become even more pertinent,
...
+In what follows, we will discuss how this works using initialization as an
+example.
 
-## Instantiating a
-Network
+## Instantiating a Network
 
 Let's see what happens when we instantiate a network. We start with our
-trusty MLP as before.
+trusty multilayer perceptron as before.
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: done

File: python/guide/module/gluon/init.md [hunk 14]
```diff
@@ -54,10 +53,16 @@ print(net.collect_params)
 print(net.collect_params())
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: removed first one

File: python/guide/module/gluon/init.md [hunk 15]
```diff
@@ -74,21 +79,28 @@ net(x)            # Forward computation.
...
-works in theory, let's see when the initialization is actually triggered. In
-order to do so, we mock up an initializer which does nothing but report a debug
-message stating when it was invoked and with which paramers.
+Now that we know how it works in theory, let's see when the initialization is
+actually triggered. In order to do so, we mock up an initializer which does
+nothing but report a debug message stating when it was invoked and with which
+paramers.
 
 ```{.python .input  n=22}
 class MyInit(init.Initializer):
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: This one is a longer comment on its own line meaning the period should be there.

File: python/guide/module/gluon/init.md [hunk 16]
```diff
@@ -21,15 +18,17 @@ convolutional networks and images this problem will become even more pertinent,
...
-## Instantiating a
-Network
+## Instantiating a Network
 
 Let's see what happens when we instantiate a network. We start with our
-trusty MLP as before.
+trusty multilayer perceptron as before.
 
 ```{.python .input}
 from mxnet import init, nd
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: Done in follow up PR.

File: python/guide/module/gluon/init.md [hunk 17]
```diff
@@ -74,21 +79,28 @@ net(x)            # Forward computation.
...
+
+Dimensional inference works like this: as soon as we knew the input
 dimensionality, $\mathbf{x} \in \mathbb{R}^{20}$ it was possible to define the
 weight matrix for the first layer, i.e. $\mathbf{W}_1 \in \mathbb{R}^{256 \times
 20}$. With that out of the way, we can progress to the second layer, define its
 dimensionality to be $10 \times 256$ and so on through the computational graph
-and bind all the dimensions as they become available. Once this is known, we can
-proceed by initializing parameters. This is the solution to the three problems
-outlined above.
+and bind all the dimensions as they become available.
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: Done in follow up PR.

File: python/guide/module/gluon/init.md [hunk 18]
```diff
@@ -117,8 +129,7 @@ the shape of the weight parameters of all layers based on the shape of the
 input. Once the system has created these parameters, it calls the `MyInit`
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: Done in follow up PR.

File: python/guide/module/gluon/init.md [hunk 19]
```diff
@@ -132,16 +143,14 @@ cause confusion. Before the first forward calculation, we were unable to
...
-can occur in two cases:
+Deferred initialization does not occur if the system knows the shape of all
+parameters when calling the `initialize` function. This can occur in two cases:
 
-* We've already seen some data and we just want to
-reset the parameters.
-* We specificed all input and output dimensions of the
-network when defining it.
+* We've already seen some data and we just want to reset the parameters.
+* We specificed all input and output dimensions of the network when defining it.
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: Done in follow up PR.

File: python/guide/module/gluon/init.md [hunk 20]
```diff
@@ -161,3 +170,127 @@ net.add(nn.Dense(10, in_units=256))
...
+changing the initializer to `Constant(1)`.
+
+```{.python .input  n=10}
+net.initialize(init=init.Constant(1), force_reinit=True)
+net[0].weight.data()[0]
+```
+
+If we want to initialize only a specific parameter in a different manner, we
+can simply set the initializer only for the appropriate subblock (or parameter)
+for that matter. For instance, below we initialize the second layer to a
```
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: This section was removed

File: python/guide/module/gluon/init.md (hunk 3 above)
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: This section was removed

File: python/guide/module/gluon/init.md (hunk 4 above)
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: This section was removed

File: python/guide/module/gluon/init.md (hunk 8 above)
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: This section was removed

File: python/guide/module/gluon/init.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: Done

File: python/guide/module/gluon/init.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: Done

File: python/guide/module/gluon/init.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: Done

File: python/guide/module/gluon/init.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: Reworded

File: python/guide/module/gluon/init.md (hunk 3 above)
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: Simplified

File: python/guide/module/gluon/init.md (hunk 8 above)
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: Chatted offline, this is ok as is

Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: I'll add a clarification. Done.

File: python/guide/module/gluon/init.md (hunk 9 above)
Comment (vishaalkapoor) [reply python/guide/module/gluon/init.md]: Done

**Comment by vishaalkapoor** [comment]
@thomelane Can you take a look then I can merge into incubator?

File: python/guide/module/gluon/init.md [hunk 21]
```diff
@@ -202,27 +200,32 @@ net[0].weight.data()[0]
...
 
 Sometimes, the initialization methods we need are not provided in the `init`
 module. At this point, we can implement a subclass of the `Initializer` class
 so that we can use it like any other initialization method. Usually, we only
-need to implement the `_init_weight` function and modify the incoming `NDArray`
-according to the initial result. In the example below, we  pick a decidedly
-bizarre and nontrivial distribution, just to prove the point. We draw the
-coefficients from the following distribution:
+need to implement the `_init_weight` function to suit our needs. In the example
+below, we  pick a decidedly bizarre and nontrivial distribution, just to prove
```
Comment (thomelane) [inline python/guide/module/gluon/init.md]: we  pick -> we pick

File: python/guide/module/gluon/init.md [hunk 22]
```diff
@@ -288,9 +291,9 @@ print(net[1].weight.data()[0] == net[2].weight.data()[0])
 ```
 
 The above example shows that the parameters of the second and third layer are
-tied. They are identical rather than just being equal. That is, by changing one
-of the parameters the other one changes, too. What happens to the gradients is
-quite ingenious. Since the model parameters contain gradients, the gradients of
-the second hidden layer and the third hidden layer are accumulated in
-`shared.params.grad()` during backpropagation.
+tied. As Python objects, they are identical rather than just being equal.
+That is, by changing one of the parameters the other one changes, too. What
```
Comment (thomelane) [inline python/guide/module/gluon/init.md]: , too -> too

File: python/guide/module/gluon/init.md [hunk 23]
```diff
@@ -161,3 +169,131 @@ net.add(nn.Dense(10, in_units=256))
...
+print(net[1].weight.data()[0] == net[2].weight.data()[0])
+```
+
+The above example shows that the parameters of the second and third layer are
+tied. As Python objects, they are identical rather than just being equal.
+That is, by changing one of the parameters the other one changes, too. What
+happens to the gradients is quite ingenious. Since the model parameters contain
+gradients, the gradients of the second hidden layer and the third hidden layer
+are accumulated in `shared.params.grad` during backpropagation.
+
```
Comment (thomelane) [inline python/guide/module/gluon/init.md]: Can you add a short conclusion to be consistent with other tutorials.
